src/api/c/wrap.cpp

```diff
@@ -19,92 +19,91 @@ using af::dim4;
 using namespace detail;
 
 template<typename T>
-static inline af_array wrap(const af_array in, const dim_t ox, const dim_t oy,
-                            const dim_t wx, const dim_t wy, const dim_t sx,
-                            const dim_t sy, const dim_t px, const dim_t py,
-                            const bool is_column) {
-    return getHandle(
-        wrap<T>(getArray<T>(in), ox, oy, wx, wy, sx, sy, px, py, is_column));
+static inline void wrap(af_array *out, const af_array in,
+                        const dim_t ox, const dim_t oy,
+                        const dim_t wx, const dim_t wy,
+                        const dim_t sx, const dim_t sy,
+                        const dim_t px, const dim_t py,
+                        const bool is_column) {
+    wrap<T>(getArray<T>(*out), getArray<T>(in), ox, oy, wx, wy, sx, sy, px, py,
+            is_column);
 }
 
-af_err af_wrap(af_array* out, const af_array in, const dim_t ox, const dim_t oy,
-               const dim_t wx, const dim_t wy, const dim_t sx, const dim_t sy,
-               const dim_t px, const dim_t py, const bool is_column) {
+af_err af_wrap_common(af_array *out, const af_array in,
+                      const dim_t ox, const dim_t oy,
+                      const dim_t wx, const dim_t wy,
+                      const dim_t sx, const dim_t sy,
+                      const dim_t px, const dim_t py,
+                      const bool is_column, bool allocate_out) {
     try {
-        const ArrayInfo& info = getInfo(in);
-        af_dtype type         = info.getType();
-        af::dim4 idims        = info.dims();
+        ARG_ASSERT(0, out != 0);
+        ARG_ASSERT(1, in != 0);
 
-        ARG_ASSERT(2, wx > 0);
-        ARG_ASSERT(3, wx > 0);
-        ARG_ASSERT(4, sx > 0);
-        ARG_ASSERT(5, sy > 0);
+        const ArrayInfo& info  = getInfo(in);
+        const af_dtype in_type = info.getType();
+        const dim4 in_dims     = info.dims();
+        const dim4 out_dims(ox, oy, in_dims[2], in_dims[3]);
 
-        dim_t nx = (ox + 2 * px - wx) / sx + 1;
-        dim_t ny = (oy + 2 * py - wy) / sy + 1;
+        ARG_ASSERT(4, wx > 0);
```

Comment:
These indicies will be incorrect in the older versions of approx. These checks should really be in the API level functions. Unfortunately it will be error prone to duplicate this everywhere. I am not sure how to fix this cleanly.